### PR TITLE
Split oversized result uploads so wide-taxonomy batches don't get rejected

### PR DESCRIPTION
**Pull request by mihow**
## Summary

For models with very large taxonomies — for example the ~29,000-class `global_moths_2024` classifier — a single batch's result upload can reach 110–142 MB, because every detection carries full per-class `scores` and `logits` arrays. In a production deployment those uploads were rejected by the reverse proxy with HTTP 413 (Request Entity Too Large), so the worker could never record its results: the upload failed, the task was never acknowledged, it eventually exhausted its redelivery budget, and the job failed with nothing stored. Smaller-taxonomy jobs were unaffected, which is why this stayed invisible until a wide-taxonomy job ran on a dense capture set.

This change makes the worker split a batch's results across multiple smaller uploads, each kept under a configurable byte cap, so wide-taxonomy jobs complete regardless of the proxy's request-body limit. It is a client-side fix that needs no server change.

Splitting is safe because of how the API already works: a request body is a container of independent results, and the server queues and acknowledges each one by its own `reply_subject`. Which upload carried a given result makes no difference to how it is processed. If one upload fails, only the images it carried go unacknowledged and are redelivered later; results from uploads that succeeded stay recorded, and are not sent twice.

### List of Changes

| # | Change (user/operator effect) | How (implementation) |
|---|---|---|
| 1 | Result uploads for large-taxonomy models now succeed instead of being rejected for size. | New `chunk_results_by_size()` greedy byte-bounded packer in `client.py`; `post_batch_results()` serializes each result once, packs results into uploads under `max_bytes`, and posts each one. A single result that exceeds the cap on its own is sent alone and logged. |
| 2 | The size cap now bounds the bytes actually sent, rather than an undercount of them. | Sizes were measured with a compact JSON encoding, but `requests` serializes with `json.dumps` defaults, which add a space after every comma and colon. On long numeric arrays that is about a quarter of the body, so the cap was advisory rather than enforced: at the 25 MB default, a batch from the ~29k-class model packed into a 25.06 MB request. Measurement now matches the encoder, and the envelope and separator costs are derived from it instead of written out. |
| 3 | A single bad response from the server no longer strands the uploads queued behind it. | The handler caught only `requests.RequestException`, so a `ValidationError` from `AntennaResultPostResponse.model_validate()` escaped and aborted the loop. It now also catches `ValueError`, which covers both that and a non-finite score failing to encode. |
| 4 | Operators can tune the per-upload size cap per deployment. | New setting `antenna_result_post_max_bytes` → env var `AMI_ANTENNA_RESULT_POST_MAX_BYTES` (default 25 MB), threaded through `ResultPoster` into the worker. The benchmark tool takes the same setting, so a benchmark reports the request count a tuned deployment would really make. |
| 5 | Regression coverage for the splitting behaviour and its failure semantics. | `trapdata/antenna/tests/test_result_chunking.py` — 13 tests: packing under the cap, splitting a large batch, over-cap single results, upload sizes measured as `requests` encodes them, and the failure semantics described below. |

### Failure semantics

Worth stating explicitly, because the two halves sound contradictory and are not:

- **A failed upload does not stop the ones after it.** Each upload carries a different set of `reply_subject`s, so continuing acknowledges work that is already finished instead of forcing it to be redone.
- **`post_batch_results()` still returns `True` only if every upload succeeded.** A partial failure returns `False` even though some results were recorded.

Both are now pinned by tests (`TestPostBatchResultsFailureSemantics`). Note that the return value currently feeds metrics and logging only; recovery from a partial failure comes from redelivery of the unacknowledged images, not from the boolean.

### Diagnosis — why one batch reaches ~140 MB

- Each `ClassificationResponse` carries `scores` and `logits`, each an array of length = number of model classes. At ~29k classes that is ~1.1 MB per detection for those two arrays alone.
- The default batch is 24 images, and trap images routinely contain many moths each, so a batch of ~130 detections lands at roughly 120–140 MB — matching the observed rejected upload sizes.
- This is payload width, not duplication: the worker builds each upload from only the current batch's detections (`worker.py`), and the classifier/detector are reset per batch, so there is no cross-batch accumulation.

### Notes and follow-ups (out of scope for this PR)

Directions to discuss, ordered by leverage. Important context: result payloads are expected to **grow**, not shrink. `logits` are needed and will be kept, and per-crop vector embeddings are planned for each detection. That makes compressing the upload more valuable over time, and argues against tightening the proxy's body-size limit.

1. **Request gzip is the highest-leverage lever, and will matter more as payloads grow.** The worker posts raw JSON, and the API server compresses *responses* but has no request-body decompression (no `Content-Encoding: gzip` handling). The big arrays — `scores`, `logits`, and soon per-crop embeddings — are numeric and compress roughly 5–10×, so gzipping uploads could cut every request at the source. It requires the API server to accept and decompress gzipped request bodies first, and the reverse proxy to pass them through. Chunking was chosen here because it is client-side only and unblocks affected jobs now; gzip is the better long-term fix and the two compose rather than compete.
2. **Sending compact JSON would cut about 20% for free.** Encoding the body in the client and posting it as bytes, rather than handing `requests` a dict, removes the whitespace `json.dumps` adds by default. It was left out of this PR because the shared test HTTP shim forwards only the `json=` argument and would need updating alongside it.
3. **Keep generous size headroom.** Because `logits` stay and embeddings are coming, neither the per-upload cap nor the proxy body-size limit should be tightened. Chunking keeps individual uploads bounded regardless of how large a full batch's results become.
4. **`labels` is already omittable for large models.** The per-classification `labels` array is `list[str] | None` and documented as "omitted if the model has too many labels" on both the worker and server schemas, so for a 29k-class model it is most likely already dropped — a minor saving already realized. The bulk is `scores` + `logits` (and soon embeddings).

### Not yet verified

- No live-worker run. Chunking is verified by unit and integration tests against the serialization path, not against a real GPU job. This should be confirmed on a real large-taxonomy job, checking that each upload lands under the deployment's proxy limit.
- If a single image produces enough detections to exceed the cap on its own (~13+ dense detections at ~29k classes), that one result is still sent in a single upload — it cannot be split below one image without an API-contract change. The configurable cap plus a generous proxy limit covers this for now.

### Test status

`uv run python -m pytest trapdata/antenna/tests/ -q` → 26 passed. Full suite → 51 passed, 1 skipped, with `test_models.py::TestSourceImageSchema::test_url` failing on an external image host; that failure is unrelated to this branch and appears on others.

### CODE REVIEW

File: trapdata/antenna/client.py [hunk 1]
```diff
@@ -71,41 +78,133 @@ def get_jobs(
...
+                    f"({len(chunk)} results) to job {job_id}: "
+                    f"{result.results_queued} queued"
+                )
+            except requests.RequestException as e:
+                logger.error(
+                    f"Failed to post result chunk {chunk_idx + 1}/{len(chunks)} "
+                    f"to {url}: {e}"
+                )
+                all_ok = False
+    return all_ok
```
[the remark this answers is not in the document]
Comment (mihow) [reply trapdata/antenna/client.py:234]: **Claude says:** Half right, and the half that is right is now fixed in `fa120c7`.

**Correct: `ValidationError` was escaping.** `AntennaResultPostResponse.model_validate()` raises Pydantic's `ValidationError`, which is not a `requests` exception, so a response that parsed as JSON but did not match the schema escaped the handler. With results now split across several POSTs that is worse than it was before the split: the escape aborts the loop, so every chunk behind the bad one goes unsent and its images wait on redelivery for no reason.

The handler now catches `ValueError` alongside `requests.RequestException`. `ValidationError` subclasses `ValueError`, so one clause covers it, and it also covers a non-finite score failing to encode.

**Not correct: `response.json()` was already covered.** `requests.exceptions.JSONDecodeError` subclasses `RequestException` (via `InvalidJSONError`), and has since requests 2.27:

```
>>> from requests.exceptions import JSONDecodeError
>>> [c.__name__ for c in JSONDecodeError.__mro__]
["JSONDecodeError", "InvalidJSONError", "RequestException", "OSError",
 "JSONDecodeError", "ValueError", "Exception", "BaseException", "object"]
```

`pyproject.toml` pins `requests>=2.28`, so a malformed body was caught across the whole supported range.

That difference is visible in the two regression tests added for this: `test_offschema_response_is_caught_per_chunk` fails against the old narrow `except` and passes now, while `test_undecodable_response_is_caught_per_chunk` passes either way. Both are kept as pins.

**Comment by mihow** [comment]
**Claude says:** This PR had been open a couple of months collecting small fixes, so before another round of line-by-line I did a structural pass: what the change is really deciding, and what might be wrong in code the diff never touches. Most of what came out is already fixed on the branch. The rest are decisions worth making on purpose rather than by default.

## Fixed on the branch

**1. The size cap was not actually capping.** (`fa120c7`)

This is the one worth reading. The packer measured each result with a compact JSON encoding, but `requests` serializes a `json=` argument using `json.dumps` defaults, which put a space after every comma and colon. On the long numeric `scores` and `logits` arrays this PR exists to handle, that whitespace is about a quarter of the body. So the cap was an undercount of the thing it was capping. At the shipped 25 MB default with the ~29k-class model, a batch packed into a 25.06 MB request:

```
29000 classes, cap 25.0 MB | before: max body 25.06 MB (100%)  <-- over cap
                           | after:  max body 23.50 MB (94%)
```

The overshoot is small at the current default and the proxy limit leaves room, so this was very unlikely to bite in practice. It mattered because the guarantee was not sound: the margin came from headroom rather than from the cap. Sizes are now measured the way `requests` encodes them, and the envelope and separator costs are derived from the encoder rather than written out, so the two cannot drift apart again.

**2. A `ValidationError` could strand every upload behind it.** (`fa120c7`)

CodeRabbit flagged this and was half right; details and the empirical check are in [that thread](https://github.com/RolnickLab/ami-data-companion/pull/149#discussion_r3763810091). Short version: `model_validate()` raises Pydantic's `ValidationError`, which is not a `requests` exception and escaped the handler. Before chunking that failed one POST. After chunking it aborts the loop, so uploads queued behind the bad response are never attempted and their images wait on redelivery for nothing. The handler now catches `ValueError` as well, which covers it. The other half of the report, JSON decode errors, was already covered, since `requests.exceptions.JSONDecodeError` subclasses `RequestException`.

**3. The five failing `test_worker.py` tests.** (`fa120c7`)

These were not a product regression, which surprised me. The worker tests build settings as a bare `MagicMock`, so the new `antenna_result_post_max_bytes` arrived as a `MagicMock` and raised `TypeError` the moment the packer compared it against an int. That got swallowed by the broad `except Exception` in `_post_results_sync` and reported as a failed post, so the symptom was `assert 0 == 2`: results that were simply never posted. Production always gets a real int from Pydantic, so the impact was confined to the tests.

Both settings helpers now supply a real number. Worth noting the same trap had already been hit once before, which is why `localization_batch_size` carries a `# Real integer` comment. A `MagicMock` settings object silently invents a value for every new setting the worker reads, so each one of these lands as a mysterious product failure. If it happens a third time, building the double from a real `Settings` would end the category.

**4. `benchmark.py` ignored the new setting.** (`da59959`)

The kind of thing a diff review will not catch, since the file is not in the diff. It built its `ResultPoster` without the cap, so it fell back to the module default and quietly disregarded `AMI_ANTENNA_RESULT_POST_MAX_BYTES`. A benchmark against a tuned deployment would report a different request count than the worker actually makes. It now takes the setting, like the worker does. The `ResultPoster` docstring also documents `max_post_bytes` and `future_timeout`, per the other CodeRabbit note.

## Worth a decision before merge

**5. Chunking over gzip, made explicit.** The body argued gzip is "the highest-leverage lever" and that payloads will grow as embeddings arrive, then shipped chunking. I think that is the right call: chunking is client-side only and unblocks affected jobs now, whereas gzip needs the API server to decompress request bodies and the proxy to pass them through. They also compose rather than compete. But it was being made by default rather than on purpose, so the body now says so. Happy to be overruled on the ordering.

**6. The success boolean is currently decorative.** `post_batch_results()` returns `True` only when every upload lands, and that is now pinned by tests. But the value feeds metrics and logging only: `_process_job` returns `did_work` regardless, so a batch with a failed upload does not fail the job. In practice recovery comes from redelivery of the unacknowledged images, which is per-image and works. I lean towards leaving it, since redelivery is the more granular mechanism and failing the job would be coarser. Flagging it because "we rely on redelivery, not on the return value" is a design position worth holding deliberately rather than discovering later.

**7. Still no live-worker run.** Everything here is verified against the serialization path, not a GPU job. The remaining unknown is the one the PR already names: an image dense enough to exceed the cap on its own still goes out in a single upload. A real large-taxonomy job, checking each upload lands under the proxy limit, is the last gap. That is a "confirm after merge" item to me rather than a blocker, given the cap now measures honestly.

## Smaller cleanups (low priority)

- The body claimed 7 tests covering "all-chunks-must-succeed semantics", but no test actually exercised a failing chunk. That claim is now true: `TestPostBatchResultsFailureSemantics` pins that a failed upload reports failure, does not abort the uploads after it, and that an off-schema or undecodable response fails only its own upload. 13 tests in that file now.
- Sending compact JSON would cut roughly 20% off every request for free, by encoding the body in the client instead of handing `requests` a dict. Left out here because the shared test HTTP shim forwards only the `json=` argument and would need updating with it, and that file is being edited on another branch. Noted as a follow-up.

## Where it stands

`trapdata/antenna` is 26/26 locally and 0 failures in CI. The full CI run is 49 passed, 3 failed on both 3.10 and 3.12, and those three are `test_api.py::test_config_num_classification_predictions`, `test_api.py::test_logits_in_classification_response`, and `test_models.py::test_url`. All three are environmental, unrelated to this branch, and being fixed separately: PR #163 fails the identical three on its current run. None of the five `test_worker.py` failures remain.

No history was rewritten, both pushes were fast-forwards, so anything stacked on this branch is unaffected.
